### PR TITLE
Channel handshake

### LINKED ISSUE #8: Implement Channel Handshake in Link
- **I'm submitting a ...**
  [ ] bug report
  [ ] feature request
  [ ] question about the decisions made in the repository
  [ ] question about how to use this project

- **Summary**

Builds on #7 

* Implement `listPorts`, `listChannels(port?: string)` on Endpoint
* Implement `createChannel` on Link

We should be able to create multiple ics20 channels between two chains, as well as find and reuse exisiting channels. Do tests that cover error handling (eg. put an invalid version so one call will fail... how to handle cleanup?)

- **Other information** (e.g. detailed explanation, stack traces, related issues, suggestions how to fix, links for us to have context, eg. StackOverflow, personal fork, etc.)

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #8

### CODE REVIEW

File: src/lib/link.ts [hunk 1]
```diff
@@ -152,44 +152,83 @@ export class Link {
...
-  //   // CreateChannel constructs and executes channel handshake messages in order to create
-  // // OPEN channels on chainA and chainB. The function expects the channels to be successfully
-  // // opened otherwise testing will fail.
-  // func (coord *Coordinator) CreateChannel(
-  // 	chainA, chainB *TestChain,
-  // 	connA, connB *ibctesting.TestConnection,
-  // 	sourcePortID, counterpartyPortID string,
-  // 	order channeltypes.Order,
-  // ) (ibctesting.TestChannel, ibctesting.TestChannel) {
+    // init on src/A
```
Comment (willclarktech) [inline src/lib/link.ts:163]: How is it that can `src` and `A` be equated? Shouldn't this just be `src`?

**Review by willclarktech — COMMENTED** [review]
Can we rename these `Link` methods too?
- `createConnection` -> `createWithNewConnections`
- `reuseConnection` -> `createWithExistingConnections`

File: src/lib/ibcclient.ts [hunk 2]
```diff
@@ -646,6 +713,172 @@ export class IbcClient {
...
+    remotePortId: string,
+    ordering: Order,
+    connectionId: string,
+    version: string
+  ): Promise<CreateChannelResult> {
+    const senderAddress = this.senderAddress;
+    const createMsg = {
+      typeUrl: '/ibc.core.channel.v1.MsgChannelOpenInit',
+      value: MsgChannelOpenInit.fromPartial({
+        portId,
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply src/lib/ibcclient.ts:728]: I'm not a fan of mixing the internal and external name. However, this is what you do with imports all the time. And the whole industry does it. 1:0 for the the rest of the world.

The thing that really bothers me is a mix of shorthand and non-shorthand notation in one onject. But this seems to be pretty standard. In Rust, this mixed notation is even enforved by the linter. 2:0 for the the rest of the world.

I should leave, writing iOS/macOS apps in Swift.

File: package.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
     "test:lint": "eslint src --ext .ts",
     "test:prettier": "prettier \"src/**/*.ts\" --list-different",
     "test:unit": "nyc --silent ava --serial",
+    "focused-test": "run-s build && yarn test:unit ./src/lib/link.spec.ts",
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply package.json:22]: I can remove it, but I find it helpful, as most current dev is on the link.spec.ts.
I will remove it before tagging v0.1.0

File: src/lib/ibcclient.ts [hunk 4]
```diff
@@ -131,6 +158,13 @@ interface ConnectionHandshakeProof {
   consensusHeight?: Height;
 }
 
+interface ChannelHandshakeProof {
+  id: ChannelCounterparty;
+  proofHeight: Height;
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply src/lib/ibcclient.ts:163]: There are two heights, shortly below we calculate a `queryHeight`. I want to make this distinction clear. One is the height of the header we are proving against, called proofHeight in the messages. We need to query `proofHeight-1` to get data that matches the appHash of the header at `proofHeight`.

I used height before, but that was confusing

File: src/lib/ibcclient.ts [hunk 5]
```diff
@@ -131,6 +158,13 @@ interface ConnectionHandshakeProof {
   consensusHeight?: Height;
 }
 
+interface ChannelHandshakeProof {
+  id: ChannelCounterparty;
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply src/lib/ibcclient.ts:162]: Maybe rename the type. In this case, we are describing the local channel.

File: src/lib/link.ts (hunk 1 above)
Comment (ethanfrey) [reply src/lib/link.ts:163]: right you are

**Comment by ethanfrey** [comment]
Thanks for the review, did a lot of cleanup at your suggestion.

**Pull request merged**